File: day_8/8.py
def draw_race(indices, length):
    lane = "~" * length
    result = []
    for i, position in enumerate(indices):
        build_lane = " " * (len(indices) - i - 1)
        if position == 0:
            build_lane += f"{lane} /{i + 1}"
        else:
            lane_with_reno = list(lane)
            real_position = position if position > 0 else length + position
            lane_with_reno[real_position] = "r"
            build_lane += "".join(lane_with_reno) + f" /{i + 1}"
        result.append(build_lane)
    return "\n".join(result)


print(draw_race([0, 5, -3], 10))
print(draw_race([2, -1, 0, 5], 8))
print(draw_race([3, 7, -2], 12))

File: day_8/8.js
/**
 * @param {number[]} indices - The reno indices
 * @param {number} length - The length of the race
 * @returns {string} The reno race
 */
function drawRace(indices, length) {
  const result = [];
  const lane = "~".repeat(length);
  for (let i = 0; i < indices.length; i++) {
    const position = indices[i];
    let buildLine = " ".repeat(indices.length - i - 1);
    if (position === 0) {
      buildLine += `${lane} /${i + 1}`;
    } else {
      let laneWithReno = lane.split("");
      const realPosition = position < 0 ? length + position : position;
      laneWithReno[realPosition] = "r";
      buildLine += `${laneWithReno.join("")} /${i + 1}`;
    }
    result.push(buildLine);
  }
  return result.join("\n");
}

console.log(drawRace([0, 5, -3], 10));
console.log(drawRace([2, -1, 0, 5], 8));
console.log(drawRace([3, 7, -2], 12));
